orttraining/orttraining/training_ops/cuda/collective/adasum_kernels.cc

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/training_ops/cuda/collective/adasum_kernels.h"
+#include "orttraining/training_ops/cpu/controlflow/common.h"
+#include "orttraining/core/framework/communication/mpi/mpi_context.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+Status AdasumAllReduce::ComputeInternal(OpKernelContext* context) const {
+
+  int vhdd_start_level = 1;
+  if (adasum_reduce_algo_ == training::AdasumReductionType::GpuHierarchical) {
+    vhdd_start_level = training::DistributedRunContext::GetInstance().GroupSize(training::WorkerGroupType::NodeLocalDataParallel);
+  }
+  // Get tensor count
+  const int num_tensors = context->InputCount();
+  std::vector<int> tensor_element_counts;
+  std::vector<size_t> tensor_offsets;
+  std::vector<size_t> tensor_sizes;
+
+  int64_t total_recv_buffer_len = 0;
+
+  size_t size_in_bytes = 0;
+  for (int i = 0; i < num_tensors; ++i) {
+    const Tensor* x_tensor = context->Input<Tensor>(i);
+    tensor_offsets.push_back(size_in_bytes);
+
+    size_in_bytes = x_tensor->SizeInBytes();
+    total_recv_buffer_len += size_in_bytes;
+
+    tensor_sizes.push_back(size_in_bytes);
+    tensor_element_counts.push_back((int)x_tensor->Shape().Size());
+  }
```

Comment:
Is there a common place that the work here can be moved to for the two implementations?